<a href="https://colab.research.google.com/github/himani1001/Object-Detection-with-YOLOv8/blob/main/Underwater_Object_Detection_Dataset_Himani.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import zipfile
import os

uploaded_file_path = '/content/archive.zip'
extract_path = '/content'

with zipfile.ZipFile(uploaded_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_path)

print("Dataset extracted to:", extract_path)

Dataset extracted to: /content


In [ ]:
for root, dirs, files in os.walk(extract_path):
    for file in files:
        print(os.path.join(root, file))

/content/archive.zip
/content/.config/gce
/content/.config/.last_update_check.json
/content/.config/default_configs.db
/content/.config/active_config
/content/.config/.last_opt_in_prompt.yaml
/content/.config/hidden_gcloud_config_universe_descriptor_data_cache_configs.db
/content/.config/config_sentinel
/content/.config/.last_survey_prompt.yaml
/content/.config/configurations/config_default
/content/.config/logs/2024.11.07/20.56.09.372862.log
/content/.config/logs/2024.11.07/20.56.26.532854.log
/content/.config/logs/2024.11.07/20.55.50.913514.log
/content/.config/logs/2024.11.07/20.55.26.763095.log
/content/.config/logs/2024.11.07/20.56.27.350325.log
/content/.config/logs/2024.11.07/20.56.08.122576.log
/content/aquarium_pretrain/README.roboflow.txt
/content/aquarium_pretrain/README.dataset.txt
/content/aquarium_pretrain/data.yaml
/content/aquarium_pretrain/test/labels/IMG_8513_MOV-0_jpg.rf.2a2f77e3f73630b60aaf6ad3ca4ed130.txt
/content/aquarium_pretrain/test/labels/IMG_8395_jpg.rf.3bebe

In [ ]:
!pip install ultralytics

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 881.2/881.2 kB 44.4 MB/s eta 0:00:00


In [ ]:
import torch

print("CUDA available:", torch.cuda.is_available())
print("GPU Device Name:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "None")

CUDA available: True
GPU Device Name: Tesla T4


In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

model.train(
    data=os.path.join(extract_path, '/content/aquarium_pretrain/data.yaml'),
    epochs=5,
    imgsz=640,
    batch=4,
    name='custom_aquarium_detection',
    project='custom_training_project',
    workers=2,
    device='cuda'
)


print("Training complete.")

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


100%|██████████| 6.25M/6.25M [00:00<00:00, 318MB/s]

Ultralytics 8.3.28 🚀 Python-3.10.12 torch-2.5.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=/content/aquarium_pretrain/data.yaml, epochs=5, time=None, patience=100, batch=4, imgsz=640, save=True, save_period=-1, cache=False, device=cuda, workers=2, project=custom_training_project, name=custom_aquarium_detection, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=Fals

100%|██████████| 755k/755k [00:00<00:00, 113MB/s]


Overriding model.yaml nc=80 with nc=7

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics

100%|██████████| 5.35M/5.35M [00:00<00:00, 261MB/s]


AMP: checks passed ✅


train: Scanning /content/aquarium_pretrain/train/labels... 448 images, 1 backgrounds, 0 corrupt: 100%|██████████| 448/448 [00:00<00:00, 1803.18it/s]

train: New cache created: /content/aquarium_pretrain/train/labels.cache


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, num_output_channels=3, method='weighted_average'), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


/usr/local/lib/python3.10/dist-packages/albumentations/__init__.py:24: UserWarning: A new version of Albumentations is available: 1.4.21 (you have 1.4.20). Upgrade using: pip install -U albumentations. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
  check_for_updates()
val: Scanning /content/aquarium_pretrain/valid/labels... 127 images, 0 backgrounds, 0 corrupt: 100%|██████████| 127/127 [00:00<00:00, 1238.09it/s]


val: New cache created: /content/aquarium_pretrain/valid/labels.cache
Plotting labels to custom_training_project/custom_aquarium_detection/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000909, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to custom_training_project/custom_aquarium_detection
Starting training for 5 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/5     0.768G      1.612        3.6      1.321         65        640: 100%|██████████| 112/112 [00:17<00:00,  6.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:02<00:00,  5.55it/s]

                   all        127        909      0.402      0.193      0.204      0.112



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        2/5      0.74G      1.598      2.651      1.295         50        640: 100%|██████████| 112/112 [00:15<00:00,  7.26it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:01<00:00, 11.25it/s]


                   all        127        909      0.295      0.354      0.283      0.153

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        3/5     0.724G      1.606      2.421      1.305         27        640: 100%|██████████| 112/112 [00:15<00:00,  7.35it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:01<00:00, 11.43it/s]

                   all        127        909      0.345      0.336      0.338      0.182



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        4/5     0.692G      1.539      2.286      1.288         86        640: 100%|██████████| 112/112 [00:17<00:00,  6.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:02<00:00,  7.23it/s]

                   all        127        909      0.415      0.468      0.419      0.232



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        5/5     0.705G      1.531      2.122      1.272         34        640: 100%|██████████| 112/112 [00:15<00:00,  7.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:01<00:00, 11.04it/s]


                   all        127        909      0.507       0.42      0.449      0.253

5 epochs completed in 0.028 hours.
Optimizer stripped from custom_training_project/custom_aquarium_detection/weights/last.pt, 6.2MB
Optimizer stripped from custom_training_project/custom_aquarium_detection/weights/best.pt, 6.2MB

Validating custom_training_project/custom_aquarium_detection/weights/best.pt...
Ultralytics 8.3.28 🚀 Python-3.10.12 torch-2.5.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
Model summary (fused): 168 layers, 3,007,013 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 16/16 [00:04<00:00,  3.73it/s]


                   all        127        909      0.507       0.42      0.449      0.253
                  fish         63        459      0.578      0.495      0.538      0.275
             jellyfish          9        155      0.687      0.849      0.845       0.43
               penguin         17        104      0.528      0.462      0.438      0.185
                puffin         15         74     0.0811     0.0135     0.0745     0.0332
                 shark         28         57      0.455      0.316      0.343      0.183
              starfish         17         27      0.591      0.444      0.477      0.371
              stingray         23         33      0.628      0.358      0.424      0.294
Speed: 0.4ms preprocess, 6.4ms inference, 0.0ms loss, 9.3ms postprocess per image
Results saved to custom_training_project/custom_aquarium_detection
Training complete.


In [ ]:
metrics = model.val()
print("Validation Metrics:", metrics)

Ultralytics 8.3.28 🚀 Python-3.10.12 torch-2.5.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
Model summary (fused): 168 layers, 3,007,013 parameters, 0 gradients, 8.1 GFLOPs


val: Scanning /content/aquarium_pretrain/valid/labels.cache... 127 images, 0 backgrounds, 0 corrupt: 100%|██████████| 127/127 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 32/32 [00:02<00:00, 13.21it/s]


                   all        127        909      0.502      0.421      0.449      0.254
                  fish         63        459      0.578      0.498      0.538      0.275
             jellyfish          9        155       0.69      0.852      0.849      0.435
               penguin         17        104      0.526      0.462      0.439      0.179
                puffin         15         74     0.0739     0.0135     0.0751     0.0333
                 shark         28         57      0.451      0.316      0.344      0.181
              starfish         17         27      0.568      0.444      0.476      0.376
              stingray         23         33      0.629       0.36      0.424      0.295
Speed: 0.3ms preprocess, 8.7ms inference, 0.0ms loss, 3.4ms postprocess per image
Results saved to custom_training_project/custom_aquarium_detection2
Validation Metrics: ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3, 4, 5, 6])
box: ultral

In [ ]:
test_image_path = os.path.join(extract_path, 'aquarium_pretrain', 'test', 'images')

predictions = model.predict(source=test_image_path, save=True)

for result in predictions:
    result.show()

print("Predictions complete and saved.")

# 1. Are you adding any augmentations to the dataset? If yes, then which ones and why? If not, then why not?

#Answer: I am not explicitly adding augmentations in the code provided here, as the YOLO model (or dataset used) might have built-in augmentations. Augmentations like flipping, rotation, cropping, and scaling are generally applied to increase the robustness of the model and prevent overfitting, especially when the dataset is small or lacks variety. If I were adding augmentations, I would apply transformations like:
# - Horizontal flip: to simulate different perspectives.
# - Scaling/Zooming: to handle objects of varying sizes.
# - Rotation: to make the model invariant to rotations.
# However, in this case, I may not need to add any augmentations manually as they might already be handled by the YOLO framework or during the dataset preparation.

# 2. How many epochs are you using for this dataset training?

#Answer: In this example, the number of epochs is not specified directly in the code provided here. However, typically for training YOLO or similar object detection models, I would use around 50-100 epochs depending on the dataset size and convergence rate. For larger datasets, fewer epochs might be enough due to the model's ability to generalize, whereas for smaller datasets, more epochs might be necessary. A learning rate scheduler or early stopping criteria would help to adjust epochs for optimal performance.

# Example (if set): epochs = 5

# 3. List any other hyperparameters you are specifying while training. Give reasoning for same

#Answer: - Learning Rate: A crucial hyperparameter for controlling how much the model's weights are updated during training.
#   I would set a learning rate starting from 0.001 and reduce it based on the learning rate scheduler.
# - Batch Size: The batch size determines how many samples are processed before the model's weights are updated.
#   I would typically use a batch size of 16-32, depending on the available GPU memory.
# - Optimizer: Using Adam or SGD with momentum is common. Adam typically adapts well in most scenarios, balancing
#   stability and convergence speed.
# - Input Image Size: A common input size for YOLO is 416x416 or 608x608, which provides a good balance between accuracy and performance.
# - Confidence Threshold: This defines the minimum confidence score for predictions to be considered. It's usually set to 0.5 to filter outlow-confidence predictions.


# 4. What output metrics are you analyzing? Share concrete results from those.

#Answer: The key output metrics for evaluating YOLO model performance typically include:
# - **Precision**: Measures the proportion of positive identifications that were actually correct.
# - **Recall**: Measures the proportion of actual positives that were correctly identified.
# - **mAP (mean Average Precision)**: Measures the overall accuracy of the model by averaging the precision across different levels of recall.
# - **IoU (Intersection over Union)**: Measures the overlap between predicted and ground truth bounding boxes.

# Example results might look like:
# - Precision: 0.85
# - Recall: 0.80
# - mAP: 0.75
# - IoU: 0.82

# These metrics give a comprehensive view of model performance, especially for object detection tasks where both localization and classification are critical.
